### PR TITLE
RavenDB-6931 assert @attachments not modified

### PR DESCRIPTION
**Pull request by talweiss1982**

### CODE REVIEW

**Review by ppekrol — APPROVED** [review]

**Comment by talweiss1982** [comment]
SlowTests.Authentication.AuthenticationLetsEncryptTests.CanGetLetsEncryptCertificateAndRenewIt failure seems unrelated

File: src/Raven.Server/Documents/DocumentCompare.cs [hunk 1]
```diff
@@ -69,19 +76,30 @@ private static DocumentCompareResult IsMetadataEqualTo(BlittableJsonReaderObject
...
                 return DocumentCompareResult.NotEqual;
             }
 
-            return ComparePropertiesExceptStartingWithAt(currentMetadata, objMetadata, true, tryMergeConflicts);
+            return ComparePropertiesExceptStartingWithAt(currentMetadata, objMetadata, true, options);
+        }
+
+        private static void ThrowAttachmentsModificationsDetected(string documentId)
+        {
+            throw new InvalidOperationException("Illegal modifications of '@attachmetns' detected for document: " + documentId);
```
Comment (ayende) [inline src/Raven.Server/Documents/DocumentCompare.cs]:
```suggestion
            throw new InvalidOperationException("Illegal modifications of '@attachments' detected for document: " + documentId);
```

File: src/Raven.Server/Documents/Patch/PatchDocumentCommand.cs [hunk 2]
```diff
@@ -182,7 +182,12 @@ protected PatchResult ExecuteOnDocument(DocumentsOperationContext context, strin
 
                             result.Status = PatchStatus.Created;
                         }
-                        else if (DocumentCompare.IsEqualTo(originalDoc, modifiedDocument, tryMergeMetadataConflicts: true) != DocumentCompareResult.Equal)
+                        else if (DocumentCompare.IsEqualTo(originalDoc, modifiedDocument, 
+                                    new DocumentCompare.DocumentCompareOptions
```
Comment (ayende) [inline src/Raven.Server/Documents/Patch/PatchDocumentCommand.cs]: Let's make this into a constant, like the `Default` field there

File: src/Raven.Server/Documents/Replication/ConflictManager.cs [hunk 3]
```diff
@@ -199,7 +199,7 @@ private static void InvalidConflictWhenThereIsNone(string id)
 
             if (existingDoc != null)
             {
-                var compareResult = DocumentCompare.IsEqualTo(existingDoc.Data, incomingDoc, true);
+                var compareResult = DocumentCompare.IsEqualTo(existingDoc.Data, incomingDoc, new DocumentCompare.DocumentCompareOptions{TryMergeMetadataConflicts = true});
```
Comment (ayende) [inline src/Raven.Server/Documents/Replication/ConflictManager.cs]: Let's make this into a constant, too

File: src/Raven.Server/Documents/DocumentCompare.cs [hunk 4]
```diff
@@ -11,8 +11,15 @@ namespace Raven.Server.Documents
 {
     public static class DocumentCompare
     {
+        public struct DocumentCompareOptions
+        {
+            public bool TryMergeMetadataConflicts;
+            public bool ThrowOnAttachmentModifications;
+            public string DocumentId;
```
Comment (ayende) [inline src/Raven.Server/Documents/DocumentCompare.cs]: Am I missing something, or is this never set?

**Review by ayende — CHANGES_REQUESTED** [review]

File: src/Raven.Server/Documents/Patch/PatchDocumentCommand.cs (hunk 2 above)
Comment (talweiss1982) [reply src/Raven.Server/Documents/Patch/PatchDocumentCommand.cs]: done

File: src/Raven.Server/Documents/Replication/ConflictManager.cs (hunk 3 above)
Comment (talweiss1982) [reply src/Raven.Server/Documents/Replication/ConflictManager.cs]: done

File: src/Raven.Server/Documents/DocumentCompare.cs (hunk 4 above)
Comment (talweiss1982) [reply src/Raven.Server/Documents/DocumentCompare.cs]: I forgot to set it but if we will use this than we can't use constant options so its allocation vs better erroring, i think it is best to catch the exception on patch only and add a better message there

File: src/Raven.Server/Documents/Patch/PatchDocumentCommand.cs [hunk 5]
```diff
@@ -182,17 +182,30 @@ protected PatchResult ExecuteOnDocument(DocumentsOperationContext context, strin
...
+                            try
+                            {
+                                compareResult = DocumentCompare.IsEqualTo(originalDoc, modifiedDocument,
+                                    DocumentCompare.DocumentCompareOptions.MergeMetadataAndThrowOnAttachmentModification);
+                            }
+                            catch (InvalidOperationException ioe)
                             {
-                                putResult = _database.DocumentsStorage.Put(context, originalDocument.Id,
-                                    originalDocument.ChangeVector, modifiedDocument, null, null, originalDocument.Flags);
+                                throw new InvalidOperationException("Illegal modifications of '@attachments' detected for document: " + id);
```
Comment (ayende) [inline src/Raven.Server/Documents/Patch/PatchDocumentCommand.cs]: Better to just state that there was an error processing this and use the `ioe` as inner exception

**Review by ayende — CHANGES_REQUESTED** [review]

File: src/Raven.Server/Documents/DocumentCompare.cs (hunk 4 above)
Comment (ayende) [reply src/Raven.Server/Documents/DocumentCompare.cs]: Sounds good, yes

File: src/Raven.Server/Documents/Patch/PatchDocumentCommand.cs (hunk 5 above)
Comment (ppekrol) [reply src/Raven.Server/Documents/Patch/PatchDocumentCommand.cs]: Done

**Pull request merged**